core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -163,21 +180,86 @@ private TableMetadata internalApply() {
     this.base = ops.refresh();
 
     Set<Long> idsToRetain = Sets.newHashSet();
-    List<Long> ancestorIds = SnapshotUtil.ancestorIds(base.currentSnapshot(), base::snapshot);
-    if (minNumSnapshots >= ancestorIds.size()) {
+    // Identify refs that should be removed
+    Set<String> refsToRemove = Sets.newHashSet();
+    for (String name : base.refs().keySet()) {
+      SnapshotRef ref = base.ref(name);
+      if (!name.equals(SnapshotRef.MAIN_BRANCH) && ref.maxRefAgeMs() != null) {
+        Snapshot snapshot = base.snapshot(ref.snapshotId());
+        if (removalTimeReference - snapshot.timestampMillis() >= ref.maxRefAgeMs()) {
+          refsToRemove.add(name);
+        }
+
+      }
+
+      if (!refsToRemove.contains(name) && !idsToRemove.contains(ref.snapshotId())) {
+        idsToRetain.add(ref.snapshotId());
+      }
+
+    }
+    List<Map.Entry<String, SnapshotRef>> branches = base
+        .refs()
+        .entrySet()
+        .stream()
+        .filter(refEntry -> refEntry.getValue().isBranch())
+        .collect(Collectors.toList());
+
+    for (Map.Entry<String, SnapshotRef> branchEntry : branches) {
+      String name = branchEntry.getKey();
+      SnapshotRef branch = branchEntry.getValue();
+      Long expireSnapshotsOlderThan = null;
+      Integer minSnapshotsToKeep = null;
+      if (name.equals(SnapshotRef.MAIN_BRANCH)) {
+        expireSnapshotsOlderThan = this.expireOlderThan;
+        minSnapshotsToKeep = this.minNumSnapshots;
+      } else if (branch.maxSnapshotAgeMs() != null) {
```

Comment:
Probably for better readability I should break this up into smaller functions

1.) Identifying refs to remove
2.) Identifying branch snapshots to keep
3.) The builder updates